From ebf9bf50fe6dd580dfbf84ca6aaaca39e493fee0 Mon Sep 17 00:00:00 2001
From: Stefan Chulski <stefanc@marvell.com>
Date: Tue, 25 Oct 2016 12:01:01 +0300
Subject: [PATCH 0553/1345] fix: mvpp2x: get virtual address of dropped
 packets using physical address

commit  90107b20c74356db846f6cac61b93ce601a22b2f from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

- we shouldn't read virtual address from descriptor since we didn't write
  virtual address during BM refill procedure, we should get virtual address
  from the physical address

Change-Id: Ib10dc92213affe3d9479de36fb7ce7c8da2736e1
Signed-off-by: Stefan Chulski <stefanc@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/33340
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Hanna Hawa <hannah@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c |   10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

diff --git a/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c b/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c
index 091d6a8..f5d2119 100644
--- a/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c
+++ b/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c
@@ -1125,13 +1125,13 @@ static void mv_pp2x_rxq_drop_pkts(struct mv_pp2x_port *port,
 		struct mv_pp2x_rx_desc *rx_desc =
 			mv_pp2x_rxq_next_desc_get(rxq);
 
-		if (port->priv->pp2_version == PPV21) {
-			buf_cookie = mv_pp21_rxdesc_cookie_get(rx_desc);
+		if (port->priv->pp2_version == PPV21)
 			buf_phys_addr = mv_pp21_rxdesc_phys_addr_get(rx_desc);
-		} else {
-			buf_cookie = mv_pp22_rxdesc_cookie_get(rx_desc);
+		else
 			buf_phys_addr = mv_pp22_rxdesc_phys_addr_get(rx_desc);
-		}
+
+		buf_cookie = phys_to_virt(dma_to_phys(port->dev->dev.parent, buf_phys_addr));
+
 		mv_pp2x_pool_refill(port->priv, MVPP2_RX_DESC_POOL(rx_desc),
 			buf_phys_addr, cpu);
 	}
-- 
1.7.9.5

